### PR TITLE
Fixed recursive variable replacement

### PR DESCRIPTION
**Pull request by noseglasses**
Fixed some problems with recursive replacement of properties.

### CODE REVIEW

File: cmake/Platform/Core/BoardPropertiesReader.cmake [hunk 1]
```diff
@@ -47,24 +47,28 @@ function(_recursively_replace_properties BOARD_ID PROPERTY_VALUE_VAR)
    # something like ${foo} is not matched as it could be a shell variable 
    # or a cmake variable or whatever, but not a Arduino property.
    #
-   while("${${PROPERTY_VALUE_VAR}}" MATCHES "[^\$]{([^}]*)}")
-      set(variable "${CMAKE_MATCH_1}")
+   while("${${PROPERTY_VALUE_VAR}}" MATCHES "(^|[^\$]){([^}]*)}")
```
Comment (noseglasses) [inline cmake/Platform/Core/BoardPropertiesReader.cmake:50]: The regex failed when variables occured directly at the start of the value string.

File: cmake/Platform/Core/BoardPropertiesReader.cmake [hunk 2]
```diff
@@ -47,24 +47,28 @@ function(_recursively_replace_properties BOARD_ID PROPERTY_VALUE_VAR)
...
       #
-      if("${variable}" MATCHES "${BOARD_ID}\.([^}]*)")
-         _get_board_property(${BOARD_ID} ${CMAKE_MATCH_1} repl_string)
-      elseif(NOT "${${variable}}" STREQUAL "")
-         # If it's not a board property, we try to find the variable 
-         # at global scope.
-         #
-         set(repl_string "${${variable}}")
-      endif()
+      _try_get_board_property("${BOARD_ID}" "${variable}" repl_string)
```
Comment (noseglasses) [inline cmake/Platform/Core/BoardPropertiesReader.cmake:57]: We are trying to find the respective variable as defined by boards.txt first.

File: cmake/Platform/Core/BoardPropertiesReader.cmake [hunk 3]
```diff
@@ -47,24 +47,28 @@ function(_recursively_replace_properties BOARD_ID PROPERTY_VALUE_VAR)
...
       
-      if(NOT "${repl_string}" STREQUAL "")
-         string(REGEX REPLACE "{${variable}}" "${repl_string}" ${PROPERTY_VALUE_VAR} "${${PROPERTY_VALUE_VAR}}")
+      if("${repl_string}" STREQUAL "")
```
Comment (noseglasses) [inline cmake/Platform/Core/BoardPropertiesReader.cmake:59]: Only if that fails, we check it the variable is defined at global scope, e.g. in platforms.txt.

File: cmake/Platform/Core/BoardPropertiesReader.cmake [hunk 4]
```diff
@@ -127,5 +131,6 @@ function(_try_get_board_property BOARD_ID PROPERTY_NAME OUTPUT_VAR)
             set(VALUE ${${BOARD_NAME}.menu.cpu.${BOARD_CPU}.${PROPERTY_NAME}})
         endif()
     endif()
+    _recursively_replace_properties(${BOARD_ID} VALUE)
```
Comment (noseglasses) [inline cmake/Platform/Core/BoardPropertiesReader.cmake]: This was already added to `_GET_BOARD_PROPERTY`. Back then I forgot to add it to `_try_get_board_property` as well.

File: cmake/Platform/Core/BoardPropertiesReader.cmake [hunk 5]
```diff
@@ -47,24 +47,28 @@ function(_recursively_replace_properties BOARD_ID PROPERTY_VALUE_VAR)
...
-      if(NOT "${repl_string}" STREQUAL "")
-         string(REGEX REPLACE "{${variable}}" "${repl_string}" ${PROPERTY_VALUE_VAR} "${${PROPERTY_VALUE_VAR}}")
+      if("${repl_string}" STREQUAL "")
+         if(NOT "${${variable}}" STREQUAL "")
+            # If it's not a board property, we try to find the variable 
+            # at global scope.
+            #
+            set(repl_string "${${variable}}")
+         else()
+            message(SEND_ERROR "Variable ${variable} used in board property definition is undefined.")
```
Comment (noseglasses) [inline cmake/Platform/Core/BoardPropertiesReader.cmake]: Have a verbose error message for the case where recursive replacement fails due to a property not being defined.

**Comment by MrPointer** [comment]
@noseglasses Changes look great, many thanks!
I have a minor request though - Could you somehow add *example* code that makes use of this recursive properties? So that the change will always be checked by the CI providers.

File: cmake/Platform/Core/BoardPropertiesReader.cmake [hunk 6]
```diff
@@ -47,24 +47,28 @@ function(_recursively_replace_properties BOARD_ID PROPERTY_VALUE_VAR)
...
       
-      if(NOT "${repl_string}" STREQUAL "")
-         string(REGEX REPLACE "{${variable}}" "${repl_string}" ${PROPERTY_VALUE_VAR} "${${PROPERTY_VALUE_VAR}}")
+      if("${repl_string}" STREQUAL "")
+         if(NOT "${${variable}}" STREQUAL "")
+            # If it's not a board property, we try to find the variable 
+            # at global scope.
+            #
+            set(repl_string "${${variable}}")
+         elseif(ARDUINO_CMAKE_ERROR_ON_UNDEFINED_PROPERTIES)
```
Comment (noseglasses) [inline cmake/Platform/Core/BoardPropertiesReader.cmake:65]: As arduino-builder seems not to complain about undefined variables that are referenced when defining other variables, we keep silent ignoring the default behavior.

File: cmake/Platform/Core/BoardPropertiesReader.cmake [hunk 7]
```diff
@@ -96,6 +100,9 @@ function(_GET_BOARD_PROPERTY BOARD_ID PROPERTY_NAME OUTPUT_VAR)
             set(VALUE ${${BOARD_NAME}.menu.cpu.${BOARD_CPU}.${PROPERTY_NAME}})
         endif()
     endif()
+    if((NOT VALUE) AND ${PROPERTY_NAME})
```
Comment (noseglasses) [inline cmake/Platform/Core/BoardPropertiesReader.cmake:103]: This enables variables form platform.txt to be found.

**Comment by noseglasses** [comment]
As this PR is still open, I added another commit that addresses reading platform.txt variables. This is related as it affects code that is changed by the first commit added to this commit. 

Concerning the example code. I am not sure how to deal with this. What you request would belong to a regression test bench which we currently not have. It would definitely require an independent firmware build and a lot of to be added to the repo. Not sure how to deal with this.

**Comment by noseglasses** [comment]
Appveyor testing failed due to a timeout of a download. Is there any way for a PR submitter to manually restart the Appveyor test?

**Comment by MrPointer** [comment]
@noseglasses I'm afraid no, at least not one that I'm aware of.

**Comment by noseglasses** [comment]
> Concerning the example code. I am not sure how to deal with this. What you request would belong to a regression test bench which we currently not have. It would definitely require an independent firmware build and a lot of to be added to the repo. Not sure how to deal with this.

@MrPointer: Could you please comment.

**Comment by MrPointer** [comment]
@noseglasses Sorry for keeping silent for a while, didn't have any time to relate to this.

> Concerning the example code. I am not sure how to deal with this. What you request would belong to a regression test bench which we currently not have. It would definitely require an independent firmware build and a lot of to be added to the repo. Not sure how to deal with this.

I completely get what you're saying, and as a matter of fact, I also agree with you.
I've always wanted to add some more sophisticated tests to this project, however, we're currently not stable enough to allow that, and by that I mean we currently just don't have enough time to deal with writing a whole test bench.
So for now, I'll just accept the changes as they are. Feel free to merge this PR once the CI tests will re-complete.

**Review by MrPointer — APPROVED** [review]

**Comment by noseglasses** [comment]
> @noseglasses Sorry for keeping silent for a while, didn't have any time to relate to this.

You're doing a great job with this project. I am sorry, if you felt pushed.

A test bench would be really nice. 

I am currently working on a [test bench](https://github.com/CapeLeidokos/Leidokos-Testing) for the Kaleidoscope firmware. There, testing is based on running virtual firmware on a x86 system using a Python scripting interface. Neglecting the fimware execution part and concentrating on the build process alone, the rest (recursive directory structure defining tests, ...) could possibly be reused for arduino-cmake. Part of this testing system is a wrapper for arduino-cmake (https://github.com/CapeLeidokos/Leidokos-CMake.git). Its travis tests build an example firmware both with arduino-cmake and arduino-builder and then perform a binary comparison. This would also eventually help arduino-cmake to make sure that the build is fully compatible with arduino-builde.

What is not included in Leidokos-Testing (as currently not necessary) are means to test different platforms (currently only the Leonardo is used). But this setup could be extended by possibly adding hardware-specification directories (defining `boards.txt`, `platform.txt`, etc.) to each test tree node. 

My own experience with test benches is that they tend to grow rapidly, causing excessive execution times. The most difficult thing would thus be to consider complex builds, even for exotic platforms to test all features, without blowing up the build times. After all, what we want is CI regression testing.

**Comment by MrPointer** [comment]
@noseglasses Haha, no, I didn't feel pushed, just want be to active as much as possible. Thanks for the concern though!

As for the test bench, it seems like you already did a really great job with your **Leidokos** project, and it sure may be our source of inspiration!
However, I must note one thing that I didn't like and it's the fact the test's goal is to compare the binaries produced from **Arduino-CMake** to those produced from **Arduino-Builder**, as they don't necessarily have to be identical - Such coupling forces us to stick to the same output, which potentially could hurt **arduino-cmake**'s progression.
Although similar, attempting to achieve the same result, the platforms build results could be different, in this case in favor of our **arduino-cmake**

And yes, you're right, test benches tend to exaggerate very quickly, but that's a conversation for another time. 😅

**Comment by noseglasses** [comment]
> However, I must note one thing that I didn't like and it's the fact the test's goal is to compare the binaries produced from Arduino-CMake to those produced from Arduino-Builder, as they don't necessarily have to be identical - Such coupling forces us to stick to the same output, which potentially could hurt arduino-cmake's progression.

The initial reason why I started comparing binaries with Arduino-Builder is that there are some people from the Kaleidoscope project that have strong feelings against CMake. To make it easier for them, I try to assert that Arduino-CMake is reliable in building firmware that is equivalent to Arduino-Builder's.

Now, after some time went by, I have a different opinion about this type of comparison. We probably agree that Arduino-Builder currently is the reference implementation of a build system for Arduino. I therefore regard it as an important part of quality assurance to make sure that we build the same binaries. From a strict IO point of view, I would expect a unique mapping of firmware code together with the content of `boards.txt`, `platforms.txt`, ... to a firmware binary.

Currently, as far as I can see, there are a lot of properties in `platforms.txt` that Arduino-CMake does not (yet) support. Comparing all our tests to Arduino-CMake could make it easier to find deficiencies when trying to eventually support all those properties.

**Comment by MrPointer** [comment]
@noseglasses Alright y then, in that case I agree that it would be a good starting point.

**Pull request merged**